### PR TITLE
Fixes #6929 KeyValue schema supports nested generics type

### LINKED ISSUE #6929: KeyValue schema in Function Error
**Describe the bug**
Use KeyValue schema in Producer

            Producer<KeyValue<Student, Student>> producer = client.newProducer(
                    KeyValueSchema.of(Schema.JSON(Student.class), Schema.JSON(Student.class)))
                    .topic("persistent://public/default/test-kvin2")
                    .producerName("producerTest").create();

Function code:
public class PulsarFunctionTest3 implements Function<KeyValue<Student, Student>, KeyValue<Student, Student>> {

    @Override
    public KeyValue<Student, Student> process(KeyValue<Student, Student> studentKeyValue, Context context) throws Exception {
        context.getLogger().info("studentKeyValue key: {}", studentKeyValue.getKey().toString());
        context.getLogger().info("studentKeyValue: {}", studentKeyValue.toString());
        return studentKeyValue;
    }
}

Pulsar Function log shows:
java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException$IncompatibleSchemaException: Key schemas or Value schemas are different schema type, from key schema type is JSON and to key schema is BYTES, from value schema is JSON and to value schema is BYTES

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by 315157973**
Fixes #6929

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

Since the schema of the keyvalue type is generic, if nested generic parameters appear, the type will be lost, resulting in the Schema type is wrong。By default, a Byte type schema will be created

Therefore, special treatment has been done for the KeyValue type schema. When creating a schema, it will traverse the KeyValue type in the function to create the correct nested generic type of schema

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest#testPulsarFunctionSchema

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by 315157973** [comment]
> @315157973 the direction of this change looks great in general. However, I am not sure adding `Type functionGenericType` into the source and sink configs is the best solution. The schema should be inferred from the argument type in the runtime. They shouldn't be tracked in the configs. Can you find a different way to get the type at runtime?

When the parameter is passed, only the Class type string of the schema is passed, and the generic type is directly discarded. The only thing we can get later is the Class string, we can't get generics.If we pass the real type instead. However, more parameter types will be modified.

JavaInstanceRunnable initializes the schema when calling “setupInput” and “setupOutput” methods. 
We must determine the schema before producers and consumers create。（Source and Sink）
I don’t understand what stage the “runtime” refers to.

If we feel that unified processing in topicSchema is not good, how about special treatment for KeyValue after "topicSchema.getSchema" is completed. This way we don't need to change the number of parameters in topicSchema.

**Comment by 315157973** [comment]
> @315157973 the direction of this change looks great in general. However, I am not sure adding `Type functionGenericType` into the source and sink configs is the best solution. The schema should be inferred from the argument type in the runtime. They shouldn't be tracked in the configs. Can you find a different way to get the type at runtime?

Get it from the context instead

**Comment by codelipenghui** [comment]
ping @sijie Please help review this PR again.

**Comment by codelipenghui** [comment]
move to 2.7.0 first

**Comment by jiazhai** [comment]
@315157973 Thanks for the fix. Seems there is some conflict. Would you please help merge your branch with the latest master?

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Schema.java [hunk 1]
```diff
@@ -369,4 +369,35 @@ default void configureSchemaInfo(String topic, String componentName,
...
+        } else if (Double.class.equals(clazz)) {
+            return Schema.DOUBLE;
+        } else if (Date.class.equals(clazz)) {
+            return Schema.DATE;
+        } else if (Time.class.equals(clazz)) {
+            return Schema.TIME;
+        } else if (Timestamp.class.equals(clazz)) {
+            return Schema.TIMESTAMP;
+        }
+        throw new IllegalArgumentException("Schema class type is incorrect");
```
Comment (sijie) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Schema.java:401]: How does this handle a POJO class?

Comment (315157973) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Schema.java:410]: > How does this handle a POJO class?

Powerful, debug with naked eyes ！

**Comment by 315157973** [comment]
There are still some problems, I will fix it soon

**Comment by 315157973** [comment]
Through the generic parameters of function, we can parse out the Java type. If other schemas are nested inside the keyValue schema, for example: 
```KeyValueSchema.of(Schema.JSON(A.class), Schema.AVRO(B.class))```
We can't know the type of complex schema nested inside, especially when there is no schem information in the broker. I now use JSON by default.
Is there any good solution for this situation?

@sijie

**Comment by 315157973** [comment]
The internal nested schema type of keyValue can be customized by user through customSchemaInputs and customSchemaOutputs. If there is no customization, the schema defaults to the parameter type of Function, and the schema of the POJO defaults to JSON.

**Pull request closed without merging**